File: inventory_tools/inventory_tools/report/quotation_demand/quotation_demand.py
# Copyright (c) 2024, AgriTheory and contributors
# For license information, please see license.txt

import json
from itertools import groupby

import frappe
from frappe.query_builder import DocType
from frappe.utils.data import fmt_money


def execute(filters=None):
	if (filters.start_date and filters.end_date) and (filters.start_date > filters.end_date):
		frappe.throw(frappe._("Start date cannot be before end date"))
	return get_columns(), get_data(filters)


def get_data(filters):
	Quotation = DocType("Quotation")
	QuotationItem = DocType("Quotation Item")
	query = (
		frappe.qb.from_(Quotation)
		.inner_join(QuotationItem)
		.on(Quotation.name == QuotationItem.parent)
		.select(
			QuotationItem.name.as_("quotation_item"),
			Quotation.name.as_("quotation"),
			Quotation.company,
			Quotation.currency,
			Quotation.party_name.as_("customer"),
			Quotation.transaction_date,
			QuotationItem.item_code,
			QuotationItem.item_name,
			QuotationItem.qty,
			QuotationItem.uom,
			QuotationItem.warehouse,
			QuotationItem.rate,
		)
		.where(Quotation.docstatus < 2)
		.where(Quotation.quotation_to == "Customer")
		.where(
			Quotation.transaction_date[filters.start_date or "1900-01-01" : filters.en_date or "2100-12-31"]
		)
		.orderby(Quotation.party_name, Quotation.name, QuotationItem.item_name)
	)

	if filters.company:
		query = query.where(Quotation.company == filters.company)

	data = query.run(as_dict=1)

	output = []
	for customer, _rows in groupby(data, lambda x: x.get("customer")):
		rows = list(_rows)
		output.append({"customer": customer, "indent": 0})
		for r in rows:
			r.split_qty = r["qty"]
			r.price = fmt_money(r.get("rate"), 2, r.get("currency")).replace(" ", "")
			r.draft_so = frappe.db.get_value(
				"Sales Order Item",
				{"quotation_item": r.quotation_item, "docstatus": 0},
				"sum(qty) as qty",
			)
			r.draft_so = f'<span style="color: red">{r.draft_so}</span>' if r.draft_so else None
			output.append({**r, "indent": 1})
	return output


def get_columns():
	hide_company = True if len(frappe.get_all("Company")) == 1 else False
	return [
		{
			"label": "Customer",
			"fieldname": "customer",
			"fieldtype": "Link",
			"options": "Customer",
			"width": "250px",
		},
		{
			"fieldname": "quotation",
			"fieldtype": "Link",
			"options": "Quotation",
			"label": "Quotation",
			"width": "170px",
		},
		{
			"fieldname": "company",
			"label": "Company",
			"fieldtype": "Data",
			"width": "200px",
			"hidden": hide_company,
		},
		{
			"fieldname": "transaction_date",
			"label": "Date",
			"fieldtype": "Date",
			"width": "100px",
		},
		{
			"fieldname": "quotation_item",
			"fieldtype": "Data",
			"hidden": 1,
		},
		{
			"fieldname": "warehouse",
			"label": "Warehouse",
			"fieldtype": "Link",
			"options": "Warehouse",
			"width": "200px",
		},
		{
			"fieldname": "item_code",
			"label": "Item",
			"fieldtype": "Link",
			"options": "Item",
			"width": "250px",
		},
		{"fieldname": "item_name", "fieldtype": "Data", "hidden": 1},
		{
			"label": "Draft SOs",
			"fieldname": "draft_so",
			"fieldtype": "Data",
			"width": "90px",
			"align": "right",
		},
		{
			"label": "Total Selected",
			"fieldname": "total_selected",
			"fieldtype": "Data",
			"width": "90px",
			"align": "right",
		},
		{
			"label": "Qty",
			"fieldname": "qty",
			"fieldtype": "Data",
			"width": "50px",
			"align": "right",
		},
		{
			"label": "Split Qty",
			"fieldname": "split_qty",
			"fieldtype": "Data",
			"width": "70px",
			"align": "right",
		},
		{"fieldname": "currency", "fieldtype": "Link", "options": "Currency", "hidden": 1},
		{
			"label": "Price",
			"fieldname": "price",
			"fieldtype": "Data",
			"width": "90px",
			"align": "right",
		},
		{"fieldname": "rate", "fieldtype": "Data", "hidden": 1},
	]


@frappe.whitelist()
def create(company, filters, rows):
	filters = frappe._dict(json.loads(filters)) if isinstance(filters, str) else filters
	rows = [frappe._dict(r) for r in json.loads(rows)] if isinstance(rows, str) else rows
	if not rows:
		return
	counter = 0
	settings = frappe.get_doc("Inventory Tools Settings", company)
	requesting_companies = list({row.company for row in rows})
	if settings.sales_order_aggregation_company == company:
		requesting_companies = [company]

	for requesting_company in requesting_companies:
		for customer, _rows in groupby(rows, lambda x: x.get("customer")):
			rows = list(_rows)
			so = frappe.new_doc("Sales Order")
			so.transaction_date = rows[0].get("transaction_date")
			so.customer = customer
			if settings.sales_order_aggregation_company and len(requesting_companies) == 1:
				so.multi_company_sales_order = True
				so.company = settings.sales_order_aggregation_company
			else:
				so.company = requesting_company
			for row in rows:
				if not row.get("item_code"):
					continue

				if settings.sales_order_aggregation_company == so.company or so.company == row.company:
					if (
						settings.sales_order_aggregation_company == so.company
						and settings.aggregated_sales_warehouse
					):
						warehouse = settings.aggregated_sales_warehouse
					else:
						warehouse = frappe.get_value("Quotation Item", row.quotation_item, "warehouse")

					so.append(
						"items",
						{
							"item_code": row.get("item_code"),
							"item_name": row.get("item_name"),
							"delivery_date": row.get("transaction_date"),
							"uom": row.get("uom"),
							"qty": row.get("split_qty"),
							"rate": row.get("rate"),
							"warehouse": warehouse,
							"quotation_item": row.get("quotation_item"),
							"prevdoc_docname": row.get("quotation"),
						},
					)

			if so.items:
				so.save()
				counter += 1

	frappe.msgprint(frappe._(f"{counter} Sales Orders created"), alert=True, indicator="green")

File: inventory_tools/inventory_tools/report/quotation_demand/quotation_demand.js
// Copyright (c) 2024, AgriTheory and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports['Quotation Demand'] = {
	filters: [
		{
			fieldname: 'company',
			label: __('Company'),
			fieldtype: 'Link',
			options: 'Company',
		},
		{
			fieldname: 'start_date',
			label: __('Start Date'),
			fieldtype: 'Date',
		},
		{
			fieldname: 'end_date',
			label: __('End Date'),
			fieldtype: 'Date',
			default: moment(),
		},
	],
	on_report_render: reportview => {
		reportview.datatable.options.columns[9].editable = true
		reportview.render_datatable()
		// these don't seem to be working
		$(".btn-default:contains('Create Card')").addClass('hidden')
		$(".btn-default:contains('Set Chart')").addClass('hidden')
	},
	get_datatable_options(options) {
		return Object.assign(options, {
			treeView: true,
			checkedRowStatus: false,
			checkboxColumn: true,
			events: {
				onCheckRow: row => {
					update_selection(row)
				},
			},
		})
	},
	onload: reportview => {
		manage_buttons(reportview)
	},
	refresh: reportview => {
		manage_buttons(reportview)
	},
}

function manage_buttons(reportview) {
	reportview.page.add_inner_button(
		__('Create SO(s)'),
		function () {
			create()
		},
		__('Create')
	)
}

function update_selection(row) {
	if (row !== undefined && !row[7].content) {
		const toggle = frappe.query_report.datatable.rowmanager.checkMap[row[0].rowIndex]
		select_all_customer_items(row, toggle).then(() => {
			update_selected_qty()
		})
	} else {
		update_selected_qty()
	}
}

function update_selected_qty() {
	// iterate all rows for selected items
	let item_map = {}
	frappe.query_report.datatable.datamanager.data.forEach((customer_row, index) => {
		if (frappe.query_report.datatable.rowmanager.checkMap[index]) {
			console.log(customer_row)
			if (customer_row.item_code && !item_map[customer_row.item_code]) {
				item_map[customer_row.item_code] = customer_row.qty
			} else if (customer_row.item_code && item_map[customer_row.item_code]) {
				item_map[customer_row.item_code] += customer_row.qty
			}
		}
	})
	frappe.query_report.datatable.datamanager.data.forEach((customer_row, index) => {
		if (customer_row.item_code in item_map) {
			let rate = Number(String(customer_row.rate).replace(/[^0-9\.-]+/g, ''))
			let total_selected = item_map[customer_row.item_code]
			let selected_price = item_map[customer_row.item_code] * (rate || 0)
			selected_price = format_currency(selected_price, customer_row.currency, 2)
			frappe.query_report.datatable.cellmanager.updateCell(9, index, total_selected, true)
			frappe.query_report.datatable.cellmanager.updateCell(12, index, selected_price, true)
		} else {
			if (customer_row.quotation) {
				// don't update indent 0 rows
				let rate = Number(String(customer_row.rate).replace(/[^0-9\.-]+/g, ''))
				selected_price = format_currency(rate, customer_row.currency, 2)
				frappe.query_report.datatable.cellmanager.updateCell(9, index, '', true)
				frappe.query_report.datatable.cellmanager.updateCell(12, index, selected_price, true)
			}
		}
	})
}

async function select_all_customer_items(row, toggle) {
	return new Promise(resolve => {
		if (frappe.query_report.datatable.datamanager._filteredRows) {
			frappe.query_report.datatable.datamanager._filteredRows.forEach(f => {
				if (f[2].content === row[1].content) {
					frappe.query_report.datatable.rowmanager.checkMap.splice(row[0].rowIndex, 0, toggle ? 1 : 0)
					$(row[0].content).find('input').check = toggle
				} else {
					frappe.query_report.datatable.rowmanager.checkMap.splice(f[0].rowIndex, 0, 0)
				}
			})
		} else {
			frappe.query_report.datatable.datamanager.rows.forEach(f => {
				if (f[2].content === row[2].content) {
					frappe.query_report.datatable.rowmanager.checkMap.splice(row[0].rowIndex, 0, toggle ? 1 : 0)
					let input = $(frappe.query_report.datatable.rowmanager.getRow$(f[0].rowIndex)).find('input')
					if (input[0]) {
						input[0].checked = toggle
					}
				} else {
					frappe.query_report.datatable.rowmanager.checkMap.splice(f[0].rowIndex, 0, 0)
				}
			})
		}
		resolve()
	})
}

async function create() {
	let filters = frappe.query_report.get_filter_values()
	if (!filters.company) {
		company = await select_company()
	} else {
		company = filters.company
	}
	let selected_rows = frappe.query_report.datatable.rowmanager.getCheckedRows()
	let selected_items = frappe.query_report.datatable.datamanager.data.filter((row, index) => {
		return selected_rows.includes(String(index)) && row.indent == 1 ? row : false
	})

	// Update split_qty with the edited value
	let selected_raw_rows = frappe.query_report.datatable.datamanager.rows.filter((row, index) => {
		return selected_rows.includes(String(index)) && row[0]['indent'] == 1 ? row : false
	})
	for (let i = 0; i < selected_items.length; i++) {
		selected_items[i]['split_qty'] = selected_raw_rows[i][11].content
	}

	if (!selected_items.length) {
		frappe.show_alert({ message: __('Please select one or more rows.'), seconds: 5, indicator: 'red' })
	} else {
		await frappe
			.xcall('inventory_tools.inventory_tools.report.quotation_demand.quotation_demand.create', {
				company: company,
				filters: filters,
				rows: selected_items,
			})
			.then(r => {})
	}
}

async function select_company() {
	return new Promise(resolve => {
		let dialog = new frappe.ui.Dialog({
			title: __('Select a Company'),
			fields: [
				{
					fieldtype: 'Link',
					fieldname: 'company',
					label: 'Company',
					options: 'Company',
					reqd: 1,
				},
			],
			primary_action: () => {
				let values = dialog.get_values()
				dialog.hide()
				return resolve(values.company)
			},
			primary_action_label: __('Select'),
		})
		dialog.show()
		dialog.get_close_btn()
	})
}
